### ISSUE TITLE
`Can't remove document listener` error on editor close

### ISSUE THREAD

**Issue by nixel2007** [ballerina-platform/lsp4intellij#20]
**Describe the bug**

Sometimes when i close editor with file, connected to lang server, I see `Can't remove document listener` error

**To Reproduce**

Can't make a reproducable scenario, but it's something like - start idea, open file, close file :)

**Expected behavior**

No errors, yay!

**Stack trace**

```
java.lang.Throwable: Can't remove document listener (com.github.lsp4intellij.editor.listeners.DocumentListenerImpl@793a7856). Registered listeners: [com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl$1@4b9e5316, 29, 0, Multicaster, 13, 307, 74, 0, 0, 83, 48, 0, 0, [], com.intellij.openapi.editor.impl.view.LogicalPositionCache@131df56e, com.intellij.openapi.editor.impl.view.TextLayoutCache@5e8a6c24, com.intellij.openapi.editor.ex.util.LexerEditorHighlighter(org.antlr.jetbrains.adapter.lexer.AntlrLexerAdapter): '...
...text of my file...
', 
use soft wraps: false, tab width: 4, additional columns: false, update in progress: false, bulk update in progress: false, dirty: true, deferred regions: []
appliance manager state: recalculation in progress: false; event being processed: null, available width: 1755, visible width: 0, dirty: true
soft wraps mapping info: Soft wraps affected by current update: []
soft wraps: [], com.intellij.openapi.editor.impl.view.EditorSizeManager@52b777c9, com.intellij.openapi.editor.impl.CaretModelImpl@468950da, com.intellij.openapi.editor.impl.InlayModelImpl$1@4889c412, com.intellij.openapi.editor.impl.EditorImpl$EditorDocumentAdapter@33c7b1ff, Multicaster, com.intellij.openapi.fileEditor.impl.text.TextEditorComponent$MyDocumentListener@19158719, com.intellij.openapi.editor.impl.ScrollingModelImpl$1@202a61ce, com.intellij.openapi.vcs.ex.DocumentTracker$MyDocumentListener@2685988d, com.intellij.openapi.vcs.ex.PartialLocalLineStatusTracker$MyUndoDocumentListener@28c39deb]
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:134)
	at com.intellij.openapi.editor.impl.DocumentImpl.removeDocumentListener(DocumentImpl.java:994)
	at com.github.lsp4intellij.editor.EditorEventManager.removeListeners(EditorEventManager.java:993)
	at com.github.lsp4intellij.client.languageserver.wrapper.LanguageServerWrapper.lambda$disconnect$2(LanguageServerWrapper.java:351)
	at java.util.concurrent.ConcurrentHashMap.forEach(ConcurrentHashMap.java:1597)
	at com.github.lsp4intellij.client.languageserver.wrapper.LanguageServerWrapper.disconnect(LanguageServerWrapper.java:349)
	at com.github.lsp4intellij.client.languageserver.wrapper.LanguageServerWrapper.disconnect(LanguageServerWrapper.java:578)
	at com.github.lsp4intellij.IntellijLanguageClient.lambda$editorClosed$1(IntellijLanguageClient.java:185)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:314)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Desktop (please complete the following information):**
 - OS: Windows 10 Pro x64
 - Browser Chrome
 - Version 73.0.3683.86

> do you really need info about browser in this bug report? :)

**Additional context**

After reopening files I see this exceptions in idea.log:

```
com.intellij.openapi.util.TraceableDisposable$DisposalException: Editor is already disposed
	at com.intellij.openapi.util.TraceableDisposable.throwDisposalError(TraceableDisposable.java:101)
	at com.intellij.openapi.editor.impl.EditorImpl.throwDisposalError(EditorImpl.java:936)
	at com.intellij.openapi.editor.impl.view.LogicalPositionCache.checkDisposed(LogicalPositionCache.java:193)
	at com.intellij.openapi.editor.impl.view.LogicalPositionCache.getLineInfo(LogicalPositionCache.java:178)
	at com.intellij.openapi.editor.impl.view.LogicalPositionCache.logicalPositionToOffset(LogicalPositionCache.java:112)
	at com.intellij.openapi.editor.impl.view.EditorCoordinateMapper.logicalPositionToOffset(EditorCoordinateMapper.java:91)
	at com.intellij.openapi.editor.impl.view.EditorView.logicalPositionToOffset(EditorView.java:161)
	at com.intellij.openapi.editor.impl.EditorImpl.logicalPositionToOffset(EditorImpl.java:2137)
	at com.github.lsp4intellij.utils.DocumentUtils.lambda$LSPPosToOffset$4(DocumentUtils.java:120)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:959)
	at com.github.lsp4intellij.utils.ApplicationUtils.computableReadAction(ApplicationUtils.java:32)
	at com.github.lsp4intellij.utils.DocumentUtils.LSPPosToOffset(DocumentUtils.java:109)
	at com.github.lsp4intellij.contributors.inspection.LSPInspection.descriptorsForManager(LSPInspection.java:90)
	at com.github.lsp4intellij.contributors.inspection.LSPInspection.checkFile(LSPInspection.java:64)
	at com.intellij.codeInspection.LocalInspectionTool$1.visitFile(LocalInspectionTool.java:141)
	at com.intellij.extapi.psi.PsiFileBase.accept(PsiFileBase.java:70)
	at com.intellij.codeInspection.InspectionEngine.acceptElements(InspectionEngine.java:75)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.lambda$null$7(LocalInspectionsPass.java:316)
	at com.intellij.util.AstLoadingFilter.lambda$toComputable$2(AstLoadingFilter.java:168)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:126)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:115)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:110)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.lambda$visitRestElementsAndCleanup$8(LocalInspectionsPass.java:315)
	at com.intellij.concurrency.ApplierCompleter.execAndForkSubTasks(ApplierCompleter.java:133)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1163)
	at com.intellij.concurrency.ApplierCompleter.lambda$wrapInReadActionAndIndicator$1(ApplierCompleter.java:105)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:582)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:532)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:87)
	at com.intellij.concurrency.ApplierCompleter.wrapInReadActionAndIndicator(ApplierCompleter.java:116)
	at com.intellij.concurrency.ApplierCompleter.lambda$compute$0(ApplierCompleter.java:96)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:147)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:222)
	at com.intellij.concurrency.ApplierCompleter.compute(ApplierCompleter.java:96)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.pollAndExecCC(ForkJoinPool.java:1190)
	at java.util.concurrent.ForkJoinPool.helpComplete(ForkJoinPool.java:1879)
	at java.util.concurrent.ForkJoinPool.awaitJoin(ForkJoinPool.java:2045)
	at java.util.concurrent.ForkJoinTask.doJoin(ForkJoinTask.java:390)
	at java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:719)
	at java.util.concurrent.ForkJoinPool.invoke(ForkJoinPool.java:2616)
	at com.intellij.concurrency.JobLauncherImpl.invokeConcurrentlyUnderProgress(JobLauncherImpl.java:65)
	at com.intellij.concurrency.JobLauncher.invokeConcurrentlyUnderProgress(JobLauncher.java:56)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.visitRestElementsAndCleanup(LocalInspectionsPass.java:328)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.inspect(LocalInspectionsPass.java:199)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.collectInformationWithProgress(LocalInspectionsPass.java:115)
	at com.intellij.codeInsight.daemon.impl.ProgressableTextEditorHighlightingPass.doCollectInformation(ProgressableTextEditorHighlightingPass.java:84)
	at com.intellij.codeHighlighting.TextEditorHighlightingPass.collectInformation(TextEditorHighlightingPass.java:69)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$null$1(PassExecutorService.java:423)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1168)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$doRun$2(PassExecutorService.java:416)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:582)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:532)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:87)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.doRun(PassExecutorService.java:415)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.lambda$run$0(PassExecutorService.java:391)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:147)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:222)
	at com.intellij.codeInsight.daemon.impl.PassExecutorService$ScheduledPass.run(PassExecutorService.java:389)
	at com.intellij.concurrency.JobLauncherImpl$VoidForkJoinTask$1.exec(JobLauncherImpl.java:161)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

**Comment by NipunaRanasinghe** [comment]
@nixel2007 Sorry for the late response since I've missed this somehow :( I could reproduce the issue on Ubuntu as well. I already fixed the `Editor is already disposed` issue and working on the document listener fix.

**Comment by nixel2007** [comment]
Still have this issue on idea-2019 branch:

```
java.lang.Throwable: Can't remove document listener (org.wso2.lsp4intellij.editor.listeners.DocumentListenerImpl@134f3b0d). Registered listeners: [com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl$1@7e1dfe3b, 4, 0, Multicaster, 0, 5, 0, 0, 0, 12, 1, 0, 0, 0, [], com.intellij.openapi.editor.impl.view.LogicalPositionCache@1746dcf, com.intellij.openapi.editor.impl.view.TextLayoutCache@6da6edbe, com.intellij.openapi.editor.ex.util.LexerEditorHighlighter(org.antlr.jetbrains.adapter.lexer.AntlrLexerAdapter): '
...file content...
', 
use soft wraps: false, tab width: 4, additional columns: false, update in progress: false, bulk update in progress: false, dirty: false, deferred regions: []
appliance manager state: recalculation in progress: false; event being processed: null, available width: 1249, visible width: 0, dirty: true
soft wraps mapping info: Soft wraps affected by current update: []
soft wraps: [], com.intellij.openapi.editor.impl.view.EditorSizeManager@706aa0c4, com.intellij.openapi.editor.impl.CaretModelImpl@19080704, com.intellij.openapi.editor.impl.InlayModelImpl$1@3d325b8f, com.intellij.openapi.editor.impl.EditorImpl$EditorDocumentAdapter@52ad0639, Multicaster, com.intellij.openapi.fileEditor.impl.text.TextEditorComponent$MyDocumentListener@38d0c832, com.intellij.openapi.editor.impl.ScrollingModelImpl$1@29ad1535, com.intellij.openapi.vcs.ex.DocumentTracker$MyDocumentListener@6ecddac]
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:145)
	at com.intellij.openapi.editor.impl.DocumentImpl.removeDocumentListener(DocumentImpl.java:977)
	at org.wso2.lsp4intellij.editor.EditorEventManager.removeListeners(EditorEventManager.java:1023)
	at org.wso2.lsp4intellij.client.languageserver.wrapper.LanguageServerWrapper.lambda$disconnect$2(LanguageServerWrapper.java:374)
	at java.util.concurrent.ConcurrentHashMap.forEach(ConcurrentHashMap.java:1597)
	at org.wso2.lsp4intellij.client.languageserver.wrapper.LanguageServerWrapper.disconnect(LanguageServerWrapper.java:372)
	at org.wso2.lsp4intellij.client.languageserver.wrapper.LanguageServerWrapper.disconnect(LanguageServerWrapper.java:611)
	at org.wso2.lsp4intellij.IntellijLanguageClient.lambda$editorClosed$1(IntellijLanguageClient.java:217)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:311)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by gayanper** [comment]
This also happens if you try to close a inactive editor while a different editor is active

**Comment by NipunaRanasinghe** [comment]
@gayanper Thanks for the clue. I'll work on the fix :)

**Comment by gayanper** [comment]
This also happened when i try to close a editor as soon as it is opened as well.

**Comment by gayanper** [comment]
I think i found the problem, 
1. its seems that all documentOpened calls are made through pooled threads. So if the opened document count exceeds the pool size they operation are waiting in the thread queue. while they are in the queue if we start closing the document this error happens. 

2. After the first problem if you did close all files, then the documentOpen requests in the queue get executed, this ends up a document which is registered with the LS while there is no document listener since we removed it earlier. Then when we try to close the last document which is now opened, it tries to remove all the listeners on connected editors on disconnect. There goes the second scenario.


In addition to this i found more concurrent issues when opening lot of files very quickly while the LS is starting up. I think we need to have a single thread which is responsible for handling the communication which we can push our requests to. Then we can control that, this communication thread will only start pumping requests once the server is initialized. @NipunaRanasinghe @rasika what do you think ?

**Comment by gayanper** [comment]
@NipunaRanasinghe @rasika have you had time to look into this ? if you guys are ok i can try to start thinking on solving this issue.

**Comment by rasika** [comment]
Didn't had time to check on this. Please go ahead, I'll also try to look into it.

**Comment by gayanper** [comment]
And also i found that when opening lot of files, if the screen space is not enough to show all file tabs, the hidden tab editors are closed by IDEA, this also cause open close synchronous issue. And found a bit questionable code at 
```java
/**
     * Disconnects an editor from the LanguageServer
     *
     * @param uri The uri of the editor
     */
    public void disconnect(String uri) {
        connectedEditors.remove(uri);
        connectedEditors.forEach((key, value) -> {
            uriToLanguageServerWrapper.remove(new ImmutablePair<>(uri, FileUtils.projectToUri(project)));
            value.removeListeners();
            value.documentClosed();
        });

        if (connectedEditors.isEmpty()) {
            stop(true);
        }
    }
```
which iterate through all the open editors and try to remove listeners. Should we only try to remove the listeners on the removed EditorEventManager ? @rasika @NipunaRanasinghe

**Comment by gayanper** [comment]
@rasika @NipunaRanasinghe please check this https://github.com/ballerina-platform/lsp4intellij/pull/75

**Comment by NipunaRanasinghe** [comment]
Closing since fixed by #75.
